### PR TITLE
Phpdoc fixes

### PR DESCRIPTION
**Pull request by garas**
Replace `@return $this` to `@return self` as recommended by PhpDoc. This fixes chained method calls in PHPStorm and possibly other IDEs.

Also update `array` to `ClassName[]` where possible, this also helps to autocomplete.

Update few incorrect datatypes.

### CODE REVIEW

**Comment by ndm2** [comment]
`self` causes problems with inheritance, when for example `B` extends `A`, and `A:foo()` returns `$this`, then the usage of `self` for `@return` will make some IDEs see `A` instead of `B`.

The problems with chaining and `$this` in PhpStorm are problems that should be fixed in PhpStorm, there are and have been several tickets regarding this - I myself have quite a few that I'm about to issue once I can find a way to reproduce them outside of CakePHP :)

**Comment by markstory** [comment]
I think giving useful docs to IDE users _now_ is helpful. We can always change the doc blocks back later when PHPStorm is fixed.

File: lib/Cake/Controller/Component/AuthComponent.php [hunk 1]
```diff
@@ -653,7 +653,7 @@ public function logout() {
  * cookies + sessions will be used.
  *
  * @param string $key field to retrieve. Leave null to get entire User record
- * @return array|null User record. or null if no user is logged in.
+ * @return mixed|null User record. or null if no user is logged in.
```
Comment (ADmad) [inline lib/Cake/Controller/Component/AuthComponent.php]: Why was this changed?

**Comment by ADmad** [comment]
Why give special preference to PHPStorm and penalize other IDE users giving incorrect results for inherited classes as @ndm2 pointed out?

**Comment by markstory** [comment]
@ADmad If this is reverted it matters not to me. I thought more people used phpstorm than other things?

**Comment by ADmad** [comment]
@markstory I don't use an IDE either so doesn't affect me. I guess we can revert if we get complains from other IDE users.

**Comment by ADmad** [comment]
From phpdoc [docs](http://www.phpdoc.org/docs/latest/guides/types.html):

> self
>    An object of the class where this type was used, if inherited it will still represent the class where it was originally defined.
> $this
>    This exact object instance, usually used to denote a fluent interface.

So just for the record using `self` is definitely not correct.

**Comment by ndm2** [comment]
@markstory That's true, however `self` will introduce other problems as its usage is incorrect, the few IDEs that handle `self` like `$this` are wrong in doing so.

I'd also like to note that most of the chaining problems left in PhpStorm, do happen when a reference like `$this` is returned with further types, like `$this|array`, when then specific inheritance/namespacing/trait-usage is involved (couldn't pin point it down yet), or when the return value is assigned to a variable, the type might get lost.

Similar problems do however occur with `self`, so it not only introduces further problems, it also doesn't fully address the ones with `$this`.

For example

```php
$email = new CakeEmail();
$email = $email
	->addBcc('')
	->addTo('')
	->attachments([]);
$email->send();
```

With both `self` and `$this`, `attachments()` and `send()` will be marked as errors, as the type is being lost. This has partially to do with overwriting the variable, using `$email2 = $email` will fix both errors for `self`, and for `$this` only the fluent interface.

There isn't much inheritance involed on the classes in the core, still I'm not sure if satisfying this rather specific problem with PhpStorm, that can relatively easily be avoided on the developer side, justifies the additional problems that `self` introduces outside of the core.

**Comment by dereuromark** [comment]
Sounds to me like phpdoc do it wrong.
$this is not a type, it is an instance of something, the correct type should actually be `self`.
This should also be respected by inheritance.

But that said, its a problem of how the phpdoc and IDE people screwed up and we have to adjust here unfortunately.
I would love to keep the intuitively "correct" version, but what does it help if it will not help you developing :/

File: lib/Cake/Controller/Component/AuthComponent.php (hunk 1 above)
Comment (garas) [inline lib/Cake/Controller/Component/AuthComponent.php]: Because expected value is not always array, often it is `string/int/bool` or other types:

I have custom function `myFunction($email)` and $email is defined as `@param string $email`.
IDE highlights `myFunction($this->Auth->user('email'))` as invalid, because it thinks I provide `array|null`, not `string`.

Similar to `Hash:get()`.

**Comment by dereuromark** [comment]
As per RFC-5: https://github.com/phpDocumentor/fig-standards/blob/master/proposed/phpdoc.md
It seems `$this` is used mainly for fluent interface and is specifically the "same instance".
`self` on the other hand should work either way and is the less strict version of those two.

**Comment by ndm2** [comment]
> $this is not a type, it is an instance of something

Technically it's both.

The thing is that there can be an actual need to be able to notate whether the return value is of the type of the current class (`self`), of the type of the last inheritor (`static`), or even the current instance (`$this`), without a substitute, `self` cannot be used for the latter. So the meaning of `self` aside, `$this` makes kinda sense, as it indicates that not just the type equals the one of current instance, but that actually the _exact same instance_ is being returned. It makes a difference whether you receive a `Foo` instance, or _the current_ `Foo` instance.

It would have been better if other words would have been chosen, as they make people think that they are ment to exactly reflect the meaning of their PHP language counterpart :/

In future this usage might become even further entrenched in case PSR-5 is being accepted in its current form, where it follows the same principle, ie `self` refers to the type where it is being used.

https://github.com/phpDocumentor/fig-standards/blob/master/proposed/phpdoc.md#keyword

**Comment by ndm2** [comment]
In case fellow PhpStorm users like to vote so things may get fixed up (faster) :)

https://youtrack.jetbrains.com/issue/WI-28091
https://youtrack.jetbrains.com/issue/WI-30131
https://youtrack.jetbrains.com/issue/WI-19953
https://youtrack.jetbrains.com/issue/WI-24959
